Context switch

magnmaeh opened the pull request:
This PR adds:

1. A very simple mechanism for setting "pins" on the FlexPRET through the emulator. It was made that simple just for testing at the time, but a better way of doing this is found in a separate PR.
2. ctx_switch.S which implements context switching. 
3. Turns interrupt_on_expire, exception_on_expire, enable_interrupts, disable_interrupts, delay_until, delay_for into macros. Also adds wait_until and wait_for macros.
4. Adds two interrupt tests for single thread and multithreaded.

Note that the target branch of this PR should be master - it is just cond-vars for now to more easily see the changes.

erlingrj commented on defines.mk:
```diff
@@ -2,9 +2,6 @@
-
 include ${FLEXPRET_ROOT_DIR}/config.mk
 
 PRINTF_DEFINES := -D PRINTF_ALIAS_STANDARD_FUNCTION_NAMES_SOFT \
```
Considering the size of printf. It would be good with information somewhere on how to disable it. Is that information brief could it maybe fit here?

magnmaeh replied:
There wasn't really an easy way to do that but I fixed it in the last commit :)

erlingrj commented on defines.mk:
```diff
@@ -14,18 +11,19 @@ PRINTF_DEFINES := -D PRINTF_ALIAS_STANDARD_FUNCTION_NAMES_SOFT \
-DEFINES += -D  DEBUG
-else ifeq ($(DEBUG), false)
-DEFINES += -D NDEBUG # Will turn assert() into nothing
+SW_DEBUG ?= true
```
Could you add brief a comment describing SW_DEBUG, DEBUG and NDEBUG

erlingrj commented on emulator/clients/Makefile:
```diff
@@ -0,0 +1,12 @@
+all: $(CLIENTS)
+
+clean:
+	rm -rf $(CLIENT_BUILD_DIR)
```
```suggestion
	rm -rf $(CLIENT_BUILD_DIR)
	
```

erlingrj replied:
All files should end in a newline: https://stackoverflow.com/a/729795

erlingrj commented on emulator/clients/common.h:
```diff
@@ -0,0 +1,64 @@
+#define PIN_IO_INT_EXTS_6 6
+#define PIN_IO_INT_EXTS_7 7
+
+#define PIN_IO_GPI_0 8
```
Is your convention that GPI_0 is a single pin, or a bank of 32 pins? It appears to be the former. But I do believe that in the CPU it is the latter?

magnmaeh replied:
I think it should be correct in the last commit :)

erlingrj commented on programs/lib/ctx_switch.S:
```diff
@@ -0,0 +1,137 @@
+ * see the macro itself for details.
+ */
+.macro getContexts
+    // Get a0 as the address of the contexts variable
```
Would be nice with a comment to understand why you need to change the assembly options before loading contexts into a0

magnmaeh replied:
Haha to be honest I don't know that myself; I just took the code from start.S. But I found an explanation that made sense online :)

erlingrj commented on programs/lib/flexpret_exceptions.c:
```diff
@@ -4,16 +4,18 @@
 static isr_t ie_int_handler;
 static isr_t ee_int_handler;
 
+struct thread_ctx_t contexts[NUM_THREADS];
```
I think you should add a little comment here telling what this variable is. Should it also be made static?

magnmaeh replied:
Making it static gives compilation errors, since the variable is referenced in the ctx_switch.S assembly code, I assume. So I guess it has to be global

erlingrj commented on programs/lib/include/flexpret_thread.h:
```diff
@@ -57,6 +57,9 @@ int tmode_sleep(uint32_t hartid);
 
 
 /* Pthreads-like threading library APIs */
```
This comment should be moved and a new comment should describe this new struct

erlingrj commented on programs/lib/include/flexpret_exceptions.h:
```diff
@@ -40,28 +40,32 @@ typedef void (*isr_t)(void);
  * 
  */
-void enable_interrupts();
+#define enable_interrupts() set_csr(CSR_STATUS, 0x10);
```
Hmmm. Normally the convention in C is that every macro is upper case. This wasnt the case when you started working on the code-base. But I am thinking we might want to change to this? Also. My preference is that the macros shouldnt include semicolons. Also to make it safe to use in any context you should wrap it in do { } while(0): https://stackoverflow.com/questions/923822/whats-the-use-of-do-while0-when-we-define-a-macro

erlingrj commented on programs/lib/include/flexpret_exceptions.h:
```diff
@@ -40,28 +40,32 @@ typedef void (*isr_t)(void);
  * 
  */
-void disable_interrupts();
+#define disable_interrupts() clear_csr(CSR_STATUS, 0x10);
```
Ref macro-discussion

erlingrj commented on programs/lib/startup.c:
```diff
@@ -77,6 +77,7 @@ void Reset_Handler() {
     // Only thread 0 performs the setup,
     // the other threads busy wait until ready.
     if (hartid == 0) {
+        
```
```suggestion
```

erlingrj commented on programs/tests/c-tests/gpio/gpio.c:
```diff
@@ -48,5 +97,10 @@ int main() {
 
     printf("Set invalid bit has no effect\n");
 
+#if HAVE_EMULATOR_CLIENT
+    run_emulator_client_tests();
+    printf("Pins were toggled in a random-ish fashion test passed\n");
+#endif // HAVE_EMULATOR_CLIENT
```
I think we should have an #error if we dont have the emulator client?

magnmaeh replied:
The reason I made it like that was to avoid having to deal with tests failing in the CICD, since it currently does not support using clients for tests. I guess this is something it should have, but I did not take the time to implement it and kept it like that instead.

But I feel like adding clients for testing with the CICD is a nice to have and could maybe be done later?

erlingrj commented on programs/tests/c-tests/gpio/gpio.c:
```diff
@@ -1,6 +1,16 @@
 #include <stdint.h>
 #include <flexpret.h>
 
+/**
```
This is connected to the pin_event server. If I remember correctly, then the GPI_1 refers to 32 pins not a single one.

erlingrj commented on programs/tests/c-tests/interrupt/interrupt.c:
```diff
@@ -0,0 +1,213 @@
+
+    printf("6th run: wait until ran sucessfully\n");
+
+#if EXTERNAL_INTERRUPT_TEST
```
Same. Test should fail if we dont run it

erlingrj commented on programs/tests/mt-c-tests/interrupt/interrupt.c:
```diff
@@ -0,0 +1,138 @@
+
+    printf("3rd run: %i threads all ran one interrupt each\n", USED_THREADS-1);
+
+#if EXTERNAL_INTERRUPT_TEST
```
Fail test here also if we dont run it completely?

erlingrj commented on programs/tests/c-tests/interrupt/interrupt.c:
```diff
@@ -0,0 +1,213 @@
+#endif // EXTERNAL_INTERRUPT_TEST
+
+    return 0;
+}
```
```suggestion
}

```

erlingrj commented on src/main/scala/Core/CSR.scala:
```diff
@@ -119,22 +119,33 @@ class CSR(implicit val conf: FlexpretConfiguration) extends Module {
     CSR_W -> io.rw.data_in
   ))
 
+  // Check whether the write does not make any alterations; i.e. it is a fake write
```
This is a good addition. Is there a simple way to add a Chisel test to verify that it works?

erlingrj commented on src/main/scala/Core/CSR.scala:
```diff
@@ -463,7 +474,7 @@ class CSR(implicit val conf: FlexpretConfiguration) extends Module {
       reg_ie1 := reg_ie
       // privileged mode with interrupts disabled
-      reg_prv := 3.U(2.W)
+      reg_prv := RegInit(VecInit(Seq.fill(conf.threads) { 3.U(2.W) }))
```
I am not sure this is what you want? A `RegInit` initializes the register once, not every time `io.exception` is true. 


Also, are all threads promoted to privileged mode when an exception occurs? (Is there a difference between exception and interrupt here?). Doesnt it make more sense to promote only a single thread? The original code writes a singular value into a vector. It should be equivalent of setting only t0 privilege to 3, and all other threads to no privilege

erlingrj replied:
When looking more into the FP code. It seems that these are execptions from CPU faults, not from IE, EE or external interrupts. 

Will an exception cause ALL threads to jump to the exception handler? Or only the thread that caused it?

magnmaeh replied:
I am not quite sure what should be the case here. All I know is that how the code used to be did not build. I just tried to understand what the intention seemed to be and tried to fix the error so it would at least build.

But I changed it away from RegInit at least

erlingrj reviewed (commented):
Very nice work, Magnus. Some nitpicks here and there. The big question is if whether your changes in the execption handling in CSR.scala is correct

magnmaeh commented:
Like we discussed last time, I added the stack guard compile option. It works in the test, but I have not benefitted from it outside the test environment yet. Further, I did some more modifications on the configuration files; being more consistent with naming swconfig/hwconfig. 

I also added an option -Werror=stack-usage=<stacksize> which lets the compile give errors if the programmer allocated too much data on the stack during compile-time. E.g., if the stacksize is 2048 and the programmer tries to allocate an array of 2100 bytes on the stack, a compile-time error is raised.

magnmaeh commented:
@erlingrj In the last commit I was able to remove tinyalloc and instead use newlib, like we discussed. It works in a threaded environment as well :)

erlingrj commented:
Thats great, do you know anything about the memory overhead?

magnmaeh commented:
The compile-time overhead from newlib's malloc implementation is 3580 bytes.  The compile-time overhead from tinyalloc implementation is 1680 bytes. So a bit more than a doubling. Finding the run-time overhead is probably a bit more difficult, but I can try to look into that as well.

Of course, newlib's implementation has the advantage of being thread-safe. I tested tinyalloc for multiple threads as well, just to be sure, and it immediately crashed.

Also, tinyalloc had the fundamental issue that it is incompatible with other newlib functionality. E.g., newlib's signal() function uses internal malloc functionality (specifically _malloc_r) that tinyalloc does not replace. This means there actually are two versions of malloc being used, and they overwrite each other's memory pools. I was able to reproduce this issue in the c-tests/malloc test case; if the tinyalloc malloc is used before signal, it could cause a crash because both tinyalloc's malloc and newlib's malloc are in play at the same time.

magnmaeh commented:
The three last commits are just that I committed code that came in handy for me when writing the project report :) Might come in handy in the future, if we want to reevaluate the performance of FlexPRET

magnmaeh commented:
The three last commits make the necessary changes to the processor so delay until and wait until can be used together with interrupts. Now it works to schedule an interrupt with ie and either run delay until or wait until and have the interrupt trigger :) For the delay until it continues delaying until its timer is finished, and wait until will exit when it gets an interrupt

magnmaeh commented:
> This looks like a very BIG PR with 94 files changed. Could this be split up into smaller chunks?

Yes, it has accumulated to a very big PR. I could probably make several merge requests, but what I would do in that case would probably be to just select a few different points in this branch's history and set up PR's there. And the PRs would be smaller but there would just be more of them.

However, I have started working on getting FlexPRET on fpga (https://github.com/pretis/flexpret/tree/fpga), and that branch is based off this context-switch branch.

Maybe another approach could be that we create a "dev" branch and instead merge this PR into there. Then I could merge this branch directly into dev and eventually merge the fpga branch into it as well. And when we finally merge dev into main, I could do the PR in a more "release" manner where I give a detailed description of all new features/fixes.

erlingrj commented:
Creating a branch called `dev` would only post-pone the problem of merging this huge PR. If it is possible to break it off into smaller pieces then it would be terrific, but it might not be so easy. You could try and see if it possible? What we want to avoid is reviewing a PR from an earlier commit that has mistakes which you have fixed in later commits.

Alternatively, a good description of the new features so we have more context when looking through the diff

magnmaeh commented on .gitignore:
Some additional changes were made to the makefile system. 

* hwconfig files are generated by scala when FlexPRET is built. These hold the configuration of FlexPRET, e.g., how much ISPM is available, how many threads are available, and so on.
* swconfig files are generated by make when software is built. These hold configurations based on choices set by software, e.g., the stacksize, debug options, whether or not to use printf, and so on.

Finally, .elf files are ignores because emulator clients (see `emulator/clients/README.md`) are compiled with the .elf extension.

magnmaeh commented on Makefrag:
```diff
@@ -26,17 +27,21 @@ OBJDUMP := $(RISCV_PREFIX)objdump
-CFLAGS = -g -static -O0 -march=rv32i -mabi=ilp32 -nostartfiles --specs=nosys.specs $(APP_DEFS) $(DEFINES)
-INCS = $(LIB_INCS) $(APP_INCS)
-LFLAGS = -T $(LINKER_SCRIPT) -L $(FP_LIB_DIR)/linker -Xlinker -Map=$(NAME).map 
+CFLAGS += -g -static -O0 -march=rv32i -mabi=ilp32 -nostartfiles --specs=nosys.specs -Werror=stack-usage=$(STACKSIZE) $(APP_DEFS) $(DEFINES)
```
`-Werror-stack-usage` throws a compile-time error if the compiler detects that the variables used in a function are too big for the stack (i.e., the STACKSIZE variable). To get this variable, the swconfig.mk is included on line 19.

magnmaeh commented on Makefrag:
```diff
@@ -26,17 +27,21 @@ OBJDUMP := $(RISCV_PREFIX)objdump
+INCS += $(LIB_INCS) $(APP_INCS)
+LFLAGS += -T $(LINKER_SCRIPT) -L $(FP_LIB_DIR)/linker -Xlinker -Map=$(NAME).map 
 
+ifeq ($(WANT_DEBUG),true)
+CFLAGS += -fstack-protector-all
+endif
```
`-f-stack-protector-all` modifies all function calls to be able to detect stack error in run-time. Doing this to all functions might be a bit much, so it is wrapped in the WANT_DEBUG if statement. 

In the normal case, `-f-stack-protector` should be used, which does this to just a select few functions. That is not the case in the code as of now though, so I'll update it.

magnmaeh commented on Makefrag:
```diff
@@ -58,12 +63,20 @@ mem: $(NAME).mem
 # runs the program (generated .mem file) using fp-emu 
-run: $(NAME).mem
-	@$(EMU) +ispm=$^;
+run: $(NAME).mem $(CLIENT)
+ifeq ($(CLIENT_CMD),)
+	@$(EMU) +ispm=$^
+else
+	@$(CLIENT_CMD) &
+	@$(EMU) +ispm=$^ --client
+	@wait
+endif
```
If the CLIENT variable is set, the emulator will be run with the `--client` argument. This intializes a server on the emulator which allows a client to connect. The client can send commands to set the pins on FlexPRET at specific clock cycles.

magnmaeh commented on emulator/clients/gpio.c:
This client is linked to a specific test in the test folder.
The client could be expanded to in the future take in a .csv file or something to set the GPIO pins.

magnmaeh commented on emulator/clients/interrupter.c:
```diff
@@ -0,0 +1,131 @@
+        switch (opt)
+        {
+        case 'a':
+            // 'Automatic' mode
```
The automatic mode is very helpful for the CICD, because the client can run without any user input. The test's makefile can specify a `CLIENT_CMD`. If this is set, the command will be run together with the emulator. 

In this way, one can write a test that expects input on pins and run it with a client that sets these pins. Since it does not require user input, it can be run in CICD.

magnmaeh commented on emulator/main.cpp:
```diff
@@ -90,17 +120,25 @@ int main(int argc, char* argv[]) {
     top->eval();
     timestamp++;
 
+    if (top->io_imem_store) {
+      printf("warn: IMEM store\n");
+      should_exit = true;
+      unknown_reason = false;
+      exit_in_n_cycles = 10;
+    }
```
What I found was that bugs that wrote to IMEM would only yield weird behaviour/crash the program at a much later point in a program. I would often spend a lot of time backtracing, just to find the original cause was IMEM writes. So this is a pragmatic way to stop the program early, since IMEM writes typically are bugs either way. However, for a bootloader, this is not a bug.

This should be incorporated into the emulator in some better way, however. Possibly find a way to make the emulator aware of whether IMEM writes should be treated as bugs or not.

magnmaeh commented on programs/lib/Makefrag-lib:
```diff
@@ -6,16 +6,25 @@ LIB_SOURCES += \
+LIB_SOURCES += $(FP_LIB_DIR)/printf/src/printf/printf.c
+LIB_INCS += 
+else
+LIB_SOURCES += $(FP_LIB_DIR)/syscalls/printf_stubs.c
```
`printf_stubs.c` just has empty implementations for the printf-familiy of functions.

magnmaeh commented on programs/lib/ctx_switch.S:
Possibly the most important part of the PR: the context switch mechanism. It required some changes in hardware to make this work.

magnmaeh replied:
Yes, I have quite a lot of unit tests for this mechanism. This file implements tests that are run in both the single threaded and multi threaded tests. At this point I've used it so much that I'm confident it works :)

https://github.com/pretis/flexpret/blob/context-switch/programs/tests/c-tests/interrupt/interrupt.c

magnmaeh commented on programs/lib/flexpret_cond.c:
I got some name conflicts when implementing multi-threaded LF for FlexPRET, so I decided to prefix some of the FlexPRET API with fp_*.

magnmaeh commented on programs/lib/flexpret_exceptions.c:
```diff
@@ -4,16 +4,27 @@
 static isr_t ie_int_handler;
 static isr_t ee_int_handler;
 
+struct thread_ctx_t contexts[NUM_THREADS];
```
This is where the registers mentioned in `programs/lib/ctx_switch.S` are stored.

magnmaeh commented on programs/lib/flexpret_exceptions.c:
```diff
@@ -4,16 +4,27 @@
+#ifndef NDEBUG
+uint32_t __stack_chk_guard = STACK_GUARD_INITVAL;
+
+FP_TEST_OVERRIDE
```
`FP_TEST_OVERRRIDE` is a macro that expands to __attribute__(("weak")) when the make variable TEST is set and to nothing otherwise. Basically, it makes it possible to override the "weak" function in a test environment. See `programs/tests/mt-c-tests/stack/stack.c` for an example on how this is used to test the stack protection.

The `FP_TEST_OVERRRIDE` macro is used to test newlib's `malloc_lock` functions as well. See `programs/tests/mt-c-tests/heap/heap.c`.

magnmaeh commented on programs/lib/flexpret_thread.c:
```diff
@@ -20,7 +22,7 @@
  */
 int slot_set(slot_t slots[], uint32_t length) {
     if (length > 8) {
-        assert(false, "No more than 8 slots supported");
+        fp_assert(false, "No more than 8 slots supported: %i given", length);
```
Assert also had name conflicts.

magnmaeh commented on programs/lib/flexpret_thread.c:
```diff
@@ -215,108 +217,140 @@ uint32_t num_threads_exited = 0;
 
 /* Pthreads-like threading library functions */
 
+static int check_args(
+    fp_thread_t *hartid,
+    void *(*start_routine)(void *)
+) {
+    if (hartid == NULL) {
+        errno = EINVAL;
+    } else if (start_routine == NULL) {
+        errno = EINVAL;
+    } else {
+        return 0;
+    }
+    return -1;
+}
```
Just moved code to reduce some duplicity.

magnmaeh commented on programs/lib/include/flexpret_cond.h:
```diff
@@ -8,25 +8,25 @@
-#define COND_INITIALIZER(lock_ptr) { .waiting = {false}, .lock = lock_ptr}
+    fp_lock_t *lock;    
+} fp_cond_t;
+#define FP_COND_INITIALIZER(lock_ptr) { .waiting = THREAD_ARRAY_INITIALIZER(false), .lock = lock_ptr}
```
From `programs/lib/include/flexpret_types.h`:

```
#if NUM_THREADS == 1
    #define THREAD_ARRAY_INITIALIZER(val) { val }
#elif NUM_THREADS == 2
    #define THREAD_ARRAY_INITIALIZER(val) { val, val }
#elif NUM_THREADS == 3
    #define THREAD_ARRAY_INITIALIZER(val) { val, val, val }
#elif NUM_THREADS == 4
    #define THREAD_ARRAY_INITIALIZER(val) { val, val, val, val }
#elif NUM_THREADS == 5
    #define THREAD_ARRAY_INITIALIZER(val) { val, val, val, val, val }
#elif NUM_THREADS == 6
    #define THREAD_ARRAY_INITIALIZER(val) { val, val, val, val, val, val }
#elif NUM_THREADS == 7
    #define THREAD_ARRAY_INITIALIZER(val) { val, val, val, val, val, val, val }
#elif NUM_THREADS == 8
    #define THREAD_ARRAY_INITIALIZER(val) { val, val, val, val, val, val, val, val }
#endif

magnmaeh commented on programs/lib/include/flexpret_csrs.h:
```diff
@@ -18,6 +17,8 @@
 #define CSR_TOHOST(tid) (CSR_TOHOST_BASE + tid)
 
 #define CSR_HWLOCK      0x520
+#define CSR_COMPARE_DU_WU 0x521
+#define CSR_COMPARE_IE_EE 0x522
```
Before this PR, delay until/wait until and interrupt on expire/exception on expire shared a timer register. That made it impossible to use DU/WU at the same time as IE/EE. We should be able to use them together, however, because it is desired to be able to schedule an interrupt in the future and sleep wait/delay until that interrupt occurs.

There are more changes in the .scala files about this.

The old CSR comment should be removed though.

magnmaeh commented on programs/lib/include/flexpret_exceptions.h:
```diff
@@ -40,28 +40,36 @@ typedef void (*isr_t)(void);
 
 /**
  * @brief Execute the interrupt on expire instruction
- * FIXME: Make into macro
```
Just fixed the fixme

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -20,8 +20,9 @@ static inline void write_tohost(uint32_t val) {
 
-#define _fp_abort(reason) do { \
-  printf("%s: %i: " PRINTF_COLOR_RED "Abort: %s\n" PRINTF_COLOR_NONE, __FILE__, __LINE__, reason); \
+#define _fp_abort(fmt, ...) do { \
+  printf("%s: %s: %i: " PRINTF_COLOR_RED "Abort:" PRINTF_COLOR_NONE, __FILE__, __func__, __LINE__); \
+  printf(fmt, ##__VA_ARGS__); \
   write_tohost(CSR_TOHOST_ABORT); \
 } while(0)
```
Same as with fp_assert(). This macro should probably not be prefixed with an underscore, however.

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -41,7 +42,7 @@ static inline void write_tohost_tid(uint32_t tid, uint32_t val) {
     case 6: write_csr(CSR_TOHOST(6), val); break;
     case 7: write_csr(CSR_TOHOST(7), val); break;
-    default: _fp_abort("Invalid thread id");
+    default: _fp_abort("Invalid thread id: ", tid);
```
```suggestion
    default: _fp_abort("Invalid thread id: %i\n", tid);
```

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -54,7 +55,7 @@ static inline void gpo_write(uint32_t port, uint32_t val) {
     case 2: write_csr(CSR_UARCH6, val); break;
     case 3: write_csr(CSR_UARCH7, val); break;
-    default: _fp_abort("Invalid port");
+    default: _fp_abort("Invalid port: ", port);
```
```suggestion
    default: _fp_abort("Invalid port: %i\n", port);
```

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -69,7 +70,7 @@ static inline void gpo_set(uint32_t port, uint32_t mask) {
     case 2: set_csr(CSR_UARCH6, mask); break;
     case 3: set_csr(CSR_UARCH7, mask); break;
-    default: _fp_abort("Invalid port");
+    default: _fp_abort("Invalid port: ", port);
```
```suggestion
    default: _fp_abort("Invalid port: %i\n", port);
```

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -85,7 +86,7 @@ static inline void gpo_clear(uint32_t port, uint32_t mask) {
     case 2: clear_csr(CSR_UARCH6, mask); break;
     case 3: clear_csr(CSR_UARCH7, mask); break;
-    default: _fp_abort("Invalid port");
+    default: _fp_abort("Invalid port: ", port);
```
```suggestion
    default: _fp_abort("Invalid port: %i\n", port);
```

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -100,7 +101,7 @@ static inline uint32_t gpo_read(uint32_t port) {
     case 2: return read_csr(CSR_UARCH6); break;
     case 3: return read_csr(CSR_UARCH7); break;
-    default: _fp_abort("Invalid port");
+    default: _fp_abort("Invalid port: ", port);
```
```suggestion
    default: _fp_abort("Invalid port: %i\n", port);
```

magnmaeh commented on programs/lib/include/flexpret_io.h:
```diff
@@ -118,7 +119,7 @@ static inline uint32_t gpi_read(uint32_t port) {
     case 2: return read_csr(CSR_UARCH2); break;
     case 3: return read_csr(CSR_UARCH3); break;
-    default: _fp_abort("Invalid port");
+    default: _fp_abort("Invalid port: ", port);
```
```suggestion
    default: _fp_abort("Invalid port: %i\n", port);
```

magnmaeh commented on programs/lib/include/flexpret_time.h:
Changed from functions to macros and added some comments.

magnmaeh commented on programs/lib/linker/flexpret.ld:
```diff
@@ -21,12 +21,18 @@
  * Taken from https://github.com/apache/nuttx/blob/master/boards/risc-v/c906/smartl-c906/scripts/ld.script
  */
-INCLUDE flexpret_config.ld
+INCLUDE flexpret_hwconfig.ld
+INCLUDE flexpret_swconfig.ld
```
These files are generated by scala (hwconfig) and make (swconfig). swconfig only contains stacksize, but could contain more in the future.

magnmaeh replied:
And hwconfig contains the other variables: DSPM_END, DSPM_START, ...

magnmaeh commented on programs/lib/linker/flexpret.ld:
```diff
@@ -71,4 +77,16 @@ SECTIONS
     } > RAM
 
     __end = ABSOLUTE(.);
+
+    .heap : ALIGN(4) {
+        __sheap = ABSOLUTE(.);
+        __eheap = ABSOLUTE(DSPM_END - TOTAL_STACKSIZES);
+    } > RAM
+
+    .stack : ALIGN(4) {
+        __estack = ABSOLUTE(.);
+        __sstack =  ABSOLUTE(. + TOTAL_STACKSIZES);
+    } > STACK
+
+    PROVIDE(__stack_start$ = __sstack);
```
These sections are added to be able to run more error handling/sanity checks in the `reset_handler` - the function that is run on boot.

magnmaeh commented on programs/lib/startup.c:
```diff
@@ -9,8 +9,6 @@
 
-#include "tinyalloc/tinyalloc.h"
-
 #define TA_MAX_HEAP_BLOCK   1000
 #define TA_ALIGNMENT        4
```
```suggestion
```

magnmaeh commented on programs/lib/syscalls/heap.c:
This file was used to debug tinyalloc not working in multi-threaded environment. It is no longer compiled and should be removed.

magnmaeh commented on programs/lib/syscalls/printf_stubs.c:
This file is compiled instead of the actual implmentations is WANT_PRINTF is set to false in the make system

magnmaeh commented on programs/lib/syscalls/syscalls.c:
```diff
@@ -128,18 +121,26 @@ int _rename (const char *oldpath, const char *newpath) {
-// FIXME: This should not be actually called because of tinyalloc.
+// Based on the implementation found here: https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=libgloss/aarch64/syscalls.c
+// with some modifications
 void *_sbrk(int incr) {
```
Had to modify this slightly to work with newlib's malloc, since malloc uses this function under the hood.

magnmaeh commented on programs/lib/syscalls/syscalls.c:
```diff
@@ -128,18 +121,26 @@ int _rename (const char *oldpath, const char *newpath) {
-   extern char   __end; /* Set by linker.  */
-   static char * heap_end;
-   char *        prev_heap_end;
+    extern char __sheap; // Set by linker; start of heap space
+    extern char __eheap; // Set by linker; max size of the heap (where stack begins)
```
Another reason as to why these variables were found in the linker script

magnmaeh commented on programs/lib/syscalls/syscalls.c:
```diff
@@ -175,3 +176,21 @@ int _gettimeofday(struct timeval *tv, void *tz) {
+
+static fp_lock_t malloc_lock = FP_LOCK_INITIALIZER;
+
+FP_TEST_OVERRIDE
+void __malloc_lock(struct _reent *r)
+{
+    fp_lock_acquire(&malloc_lock);
+}
+
+FP_TEST_OVERRIDE
+void __malloc_unlock(struct _reent *r)
+{
+    fp_lock_release(&malloc_lock);
+}
```
Newlib's malloc requires the user to implement these functions to work in a multi-threaded environment. The `FP_TEST_OVERRIDE` is used to check that malloc fails if these are implemented as doing nothing in a test.

magnmaeh commented on programs/tests/Makefile:
```diff
@@ -46,8 +56,9 @@ endif
 %.out: %.mem
-	-@set -o pipefail; $(EMU) +ispm=$< | tee $@; $(RES_PARSER) $@ $$?
-	
+	-@set -o pipefail
+	@make run -C $(dir $@) | tee $@
+	@$(RES_PARSER) $@ $$?
```
Quite cryptographic what is going on here; will add a comment in the code

magnmaeh commented on programs/tests/Makefile:
```diff
@@ -63,7 +74,7 @@ run-c-tests: $(EMU) $(C_TESTS_RES)
 
 # Remove all generated files by calling make clean in all the test dirs
-clean:
+clean::
```
Two colons makes it so the target can be "appended" later on

magnmaeh commented on programs/tests/c-tests/interrupt-delay/Makefile:
Here is an example on how a test can be initialized with a client automatically. The `CLIENT_CMD` is run together with the emulator by the `run` target if `CLIENT` is set. This makes it possible to run the emulator with a pins being set in CICD.

magnmaeh commented on programs/tests/c-tests/interrupt/interrupt.c:
I think these tests fail for some hw_thread configuration of FlexPRET. Will look into that before we merge.

erlingrj replied:
Any updates here?

magnmaeh replied:
Yes, I've been able to fix the bugs in the three last commits.

* The bug was initially fixed here: https://github.com/pretis/flexpret/pull/82/commits/2cc2acc5ba1f4102f6e761d8a95b134d993e233c
* Then support for multi-threaded interrupts was added here: https://github.com/pretis/flexpret/pull/82/commits/695ee87dc383e69da462476ceeb192410ab7a035
* Then a bug and some tests for this fix were added here: https://github.com/pretis/flexpret/pull/82/commits/cfb07dc0c78def9208301e27383485751dc1d17e

But to be clear, this review incorporates all those changes.

magnmaeh commented on programs/tests/mt-c-tests/heap/Makefile:
```diff
@@ -0,0 +1,5 @@
+FLEXPRET_ROOT_DIR = ../../../../
+NAME = heap
+APP_SOURCES = *.c
+TEST_ENVIRONMENT = true
```
`TEST_ENVIRONMENT` ends up setting `FP_TEST_OVERRIDE` to `__attribute__(("weak"))`, which allows us to override `__malloc_lock()` and `__malloc_unlock()` in the next file. When `TEST_ENVIRONMENT` is false or not set, trying to override the malloc functions will yield a compile error.

erlingrj replied:
Why do we want to override `__malloc_lock` and `__malloc_unlock`?

magnmaeh replied:
Just to be able to test it elsewhere. See the link. Here, if the user changes `#if 0` to `#if 1`, we override the implementations with no actual lock, and malloc stops working. So this shows it actually does something in the multi-threaded case.

https://github.com/pretis/flexpret/blob/context-switch/programs/tests/mt-c-tests/heap/heap.c

But when this test is not run, there is no reason for overriding these two functions. That's why the weak attribute is conditionally added as well

magnmaeh commented on programs/tests/mt-c-tests/interrupt/interrupt.c:
Should update this test to remove the `EXTERNAL_INTERRUPT_TEST` macro.

magnmaeh commented on programs/tests/mt-c-tests/stack/stack.c:
```diff
@@ -0,0 +1,108 @@
+
+static bool stack_fail_fnc_called[NUM_THREADS] = THREAD_ARRAY_INITIALIZER(false);
+
+// The function is weakly linked, so we override its implementation so it does
+// not exit and fail the test
```
```suggestion
// The function is weakly linked. Its default implementation would exit and fail
// the test. In this test, we want the function to be called but not fail the test.
// The solution is to override it with something that does not exit. 
```

magnmaeh commented on programs/tests/mt-c-tests/uart/client/stimuli.py:
A python script can be used as a client as well :)

magnmaeh commented on programs/tests/mt-c-tests/uart/uart.c:
Does not work correctly when trying to use two software UARTs simultaneously

magnmaeh commented on src/main/scala/Core/CSR.scala:
```diff
@@ -69,12 +71,15 @@ class CSR(implicit val conf: FlexpretConfiguration) extends Module {
   // timing instructions
   val reg_time = RegInit(0.U(64.W)) // RO, less bits if !conf.stats
-  val reg_compare = Reg(Vec(conf.threads, UInt(conf.timeBits.W)))
+  val reg_compare_du_wu = Reg(Vec(conf.threads, UInt(conf.timeBits.W)))
+  val reg_compare_ie_ee = Reg(Vec(conf.threads, UInt(conf.timeBits.W)))
```
Need one register for DU/WU and one for IE/EE, otherwise they will not work together

magnmaeh commented on src/main/scala/Core/CSR.scala:
```diff
@@ -119,22 +126,33 @@ class CSR(implicit val conf: FlexpretConfiguration) extends Module {
+    when((io.rw.write && !fake_write) && (addr_read_only || addr_no_priv)) {
       priv_fault := true.B
     }
-    // TODO: read without permission?
```
The above code implements this TODO, but FlexPRET is typically not compiled with `priviledgedMode`. I implemented this because I thought I needed `priviledgedMode` for context switches, but that was not the case after all

magnmaeh commented on src/main/scala/Core/CSR.scala:
```diff
@@ -35,14 +35,16 @@ class CSR(implicit val conf: FlexpretConfiguration) extends Module {
     val epc = Input(UInt(32.W)) // PC of uncommitted instruction
     val cause = Input(UInt(CAUSE_WI.W))
     val evecs = Output(Vec(conf.threads, UInt(32.W)))
+    val mepcs = Output(Vec(conf.threads, UInt(32.W)))
```
mepcs = machine exception PC (plural)
The RISC-V specification mentions three levels of exceptions (M, S, U, if I remember correctly). The M is the standard level to implement

magnmaeh commented on src/main/scala/Core/CSR.scala:
```diff
@@ -35,14 +35,16 @@ class CSR(implicit val conf: FlexpretConfiguration) extends Module {
     val dec_tid = Input(UInt(conf.threadBits.W))
     // privileged
-    val sret = Input(Bool()) // valid sret instruction (ignores exception)
+    val xret = Input(UInt(2.W))
```
X here refers to either of M, U, S. I used X instead of M to make it simpler to understand how this can be expanded in the future, but it might just be easier to name it `mret` instead.

erlingrj replied:
Why width of 2 and not a boolean as it used to be?

magnmaeh replied:
I tried to encode the three possible privilege levels RISC-V defines: `mret`, `sret`, and `uret`, and that requires two bits. We are only using `mret` however. I think we discussed this a while ago, whether it was good idea to encode the `sret` and `uret` as well when not using them. And I think we concluded that the two others should be removed from the code. So I'll change the code to remove the two not in use.

magnmaeh commented on src/main/scala/Core/Datapath.scala:
```diff
@@ -23,6 +24,7 @@ class Datapath(val debug: Boolean = false)(implicit conf: FlexpretConfiguration)
     val host = new HostIO()
     val gpio = new GPIO()
     val int_exts = Input(Vec(conf.threads, Bool()))
+    val imem_store = Output(Bool())
```
This is routed out to emulator to be able to warn about IMEM writes

magnmaeh commented on src/main/scala/Core/control.scala:
I ended up just expanding the decode table for the added control signals. Not sure if this is the optimal solution

magnmaeh reviewed (commented):
I went over all the changes and added some summarizing/explanatory comments. Will give an overview of the major changes too. So I think if you go over the files now, you should be able to find my comments which might speed up your review by a bit.

Also, something I was reminded when I reviewed the code myself is that many files are changed just because I have prefixed some of the FlexPRET library API with fp_*. E.g., `thread_create` -> `fp_thread_create`. I did this to avoid some name conflicts when porting to LF. But this also causes a lot of changed in tests that are otherwise unchanged.

magnmaeh commented:
The major features added in this PR are:

1. Modifications to the make system. There are in total six configuration files: `swconfig.mk`, `hwconfig.mk`, `flexpret_swconfig.h`, `flexpret_hwconfig.h`, `flexpret_swconfig.ld`, `flexpret_hwcondig.ld`. The five latter are .gitignored. 
    * `swconfig.mk`: Contains top-level configuration points for software that can be overridden in a program's `Makefile`. E.g., `STACKSIZE`, `WANT_DEBUG`, and so on.
    * `hwconfig.mk`: Is generated by Scala when FlexPRET is built. Contains the configuration of the currently built FlexPRET. E.g., `THREADS`, `ISPM_KBYTES`, `DSPM_KBYTES`, and so on.
    * `flexpret_swconfig.h`: Contains configuration derived from the `swconfig.mk`, but in the form of C-language macros so it can be used in C code. Is generated when software is built.
    * `flexpret_hwconfig.h`: Same as `flexpret_swconfig.h`, but for hardware. Is generated by Scala. 
    * `flexpret_swconfig.ld`: Same information as `flexpret_swconfig.h`, but in the Linker language so it can be used by linker scripts.
    * `flexpret_hwconfig.ld`: Same information as `flepret_hwconfig.h`, but in Linker language.
2. Replaced `tinyalloc` submodule with just newlib's malloc due to reasoning here: [https://github.com/pretis/flexpret/pull/82#issuecomment-1859173087](https://github.com/pretis/flexpret/pull/82/url).
3. Added more compile-time checks and run-time checks/debugging.
    * `-Werror-stack-usage` yields compile-time error if stack usage inside a single function is greater than the `STACKSIZE` variable in `swconfig.mk`.
    * `-f-stack-protector` yields run-time error if stack is broken.
    * Emulator yields error if instruction memory is written.
    * `fp_assert` now takes a condition and format list like `printf`. E.g., asserts are used link this now: `fp_assert(value == 15, "Value not as expected, was %i\n", value);`
    * Added symbols to linker script which contain information on where the heap and stack start and end. These are used to perform sanity checks on the configured stack and heap when any program initializes.
4. Client pin service
    * See `emulator/clients/README.md` for more details.
    * The client pin service can be initialized automatically for CICD purposes if the variables `CLIENT` and `CLIENT_CMD` are specified in the makefile.
5. Context switch mechanism
    * This required some hardware changes - essentially adding the `mret` instruction.
    * Also added `in_interrupt` bit to the `CSR_STATUS` register, which  is needed by LF port.
6. Decoupled `DU/WU` and `IE/EE` instructions so they can be used at the same time. They used to be implemented using the same timer.
7. Prefixed large part of FlexPRET API with `fp_*` to avoid name conflicts with other APIs (e.g., LF)

erlingrj commented on programs/lib/flexpret_exceptions.c:
```diff
@@ -41,66 +52,52 @@ static const char *exception_to_str(const uint32_t cause) {
+    // Call the function to load the thread's context
+    // We mark it with attribute noretun because the function will not return
+    // to fp_exception_handler, but instead to where the exception occurred.
+    void thread_ctx_switch_load(void) __attribute__((noreturn));
```
I havent seen this before. Is it a forward declaration of a function, within another function? Couldnt `thread_ctx_switch_load` be defined with that attribute in the original declaration?

magnmaeh replied:
There is no original declaration of the function in C because it is implemented in assembly (see ctx_switch.S)

It could of course be declared at file scope or in a header file, but since `fp_exception_handler` is the only function that uses it, I added it here. I think declaring it inside a function is just a matter of scoping - it does not make a difference to declare it at file level scope. But I guess it is a bit confusing to declare it inside the function, so I can move it outside.

The advantage of having it declared inside the function is that it becomes clear that it is not used anywhere else. A disadvantage is that it might cause some confusion. 

For the record: The `__attribute_((noreturn))` is not at all necessary; it just slightly reduces code size because it removes the "return to caller" part of the function.

erlingrj replied:
OK, I would prefer putting this at the top of the file, to avoid any confusion :)

erlingrj commented on programs/lib/flexpret_exceptions.c:
```diff
@@ -41,66 +52,52 @@ static const char *exception_to_str(const uint32_t cause) {
-    write_csr(CSR_EVEC, (uint32_t) fp_exception_handler);
+    // Register the function to call on exceptions; this function stores the
+    // thread's context and calls the fp_exception_handler function afterwards
+    void thread_ctx_switch_store(void);
```
Same as above

erlingrj commented on programs/lib/include/flexpret_assert.h:
```diff
@@ -4,42 +4,17 @@
 // idea to use the same variable
 #ifdef NDEBUG
-#define assert(cond, reason) ((void)0)
+#define fp_assert(cond, fmt, ...) ((void)0)
```
I think this might be problematic if you do something like `fp_assert(do_something() == true)`, if `NDEBUG` is defined `do_something` is optimized away. You want to do
`#define fp_assert(cond, fmt, ...) ((void) cond)` or something similar

magnmaeh replied:
That is a very good point! Will fix asap :)

erlingrj commented on programs/lib/include/flexpret_time.h:
```diff
@@ -4,31 +4,54 @@
-}
+#define fp_delay_for(ns) do { \
+    uint32_t now_ns = rdtime(); \
+    fp_delay_until(now_ns + ns); \
```
```suggestion
    fp_delay_until(now_ns + (ns)); \
```

erlingrj commented on programs/lib/include/flexpret_time.h:
```diff
@@ -4,31 +4,54 @@
+ */
+#define fp_wait_for(ns) do { \
+    uint32_t now_ns = rdtime(); \
+    fp_wait_until(now_ns + ns); \
```
```suggestion
    fp_wait_until(now_ns + (ns)); \
```

erlingrj commented on programs/lib/syscalls/heap.c:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @author Magnus Mæhlum (magnmaeh@stud.ntnu.no)
+ * @brief A wrapper module to expose standard heap functions (malloc, free, etc.),
```
I thought we were using newlib now? Here you use tinyalloc

magnmaeh replied:
The file is not compiled, I just forgot to delete it

erlingrj commented on programs/tests/mt-c-tests/uart/uart.c:
```diff
@@ -0,0 +1,236 @@
+/**
+ * @author Magnus Mæhlum (magnmaeh@stud.ntnu.no)
+ * @brief This test implements a Software-defined UART, which is a more formal
```
Very cool. The SD UART should be a library that the user can include. Look at how XMOS has done it: https://github.com/orgs/xmos/repositories?page=2

erlingrj commented on programs/tests/mt-c-tests/uart/uart.c:
```diff
@@ -0,0 +1,236 @@
+    STATE_STARTBIT,
+    STATE_DATABITS,
+    STATE_STOPBIT,
+} state = STATE_STARTBIT;
```
This is new syntax to me, what does this assignment mean?

magnmaeh replied:
It is equivalent to:

```
enum State {
    STATE_STARTBIT,
    STATE_DATABITS,
    STATE_STOPBIT,
};
enum State state = STATE_STARTBIT;
```

Just more short-hand syntax, I guess :) It can be done to structs as well, I think.

erlingrj commented on programs/tests/mt-c-tests/uart/uart.c:
```diff
@@ -0,0 +1,236 @@
+    STATE_STOPBIT,
+} state = STATE_STARTBIT;
+
+struct RingBuffer {
```
A ring-buffer is also something that would be nice to have as a lib that can be optionally included by users

erlingrj commented on src/main/scala/Core/Datapath.scala:
```diff
@@ -112,16 +120,26 @@ class Datapath(val debug: Boolean = false)(implicit conf: FlexpretConfiguration)
   // higher priority than BRJMP
   if (conf.exceptions) {
```
Below you have added an extra `when` comparing `nest_pc_sel_csr_addr`, can you explain why?

magnmaeh replied:
Are you referring to the `when(io.control.next_pc_sel_csr_addr === CSRs.mepc.U)`?

This is part of implementing the `mret` instruction, that was not available before. The `when` is only triggered when an `mret` instruction is run. This sets the next PC to the value in the `mepc`, which is the machine exception PC (i.e., the PC we had before the exception occured). That is how we return back to where exceptions occured.

erlingrj commented on src/main/scala/Core/constants.scala:
```diff
@@ -98,6 +98,13 @@ object FlexpretConstants
   val MEM_RD_MUL = 2.U(2.W)
   val MEM_RD_X   = BitPat("b??")
+  
+  // priviledge level of return
```
Can you elaborate a little more in this comment what those constants mean?

magnmaeh replied:
Yes I can, but I assume we are going to want to remove them anyway, actually. But if we keep them I can elaborate :)

erlingrj commented on src/main/scala/Core/control.scala:
```diff
@@ -39,6 +41,7 @@ class ControlDatapathIO(implicit val conf: FlexpretConfiguration) extends Bundle
 
   // outputs to datapath (control dependent)
   val next_pc_sel     = Output(Vec(conf.threads, UInt(NPC_WI.W)))
+  val next_pc_sel_csr_addr = Output(UInt(12.W))
```
What does this signal do?

magnmaeh replied:
It is for the exception/interrupt implementation. 

There are two scenarios where we need to read a CSR register into the PC:
1. When an exception/interrupt occurs and we want to jump to the handler. Then we load `PC <- CSR_EVEC`.
2. When we return from an exception/interrupt and we want to jump back to the original PC. Then we load `PC <- CSR_MEPC`. 

We could add two more "enums" to the `next_pc_sel` signal, like so:

```
// next PCs select
val NPC_WI = 3
val NPC_PCREG = 0.U(3.W)
val NPC_PLUS4 = 1.U(3.W)
val NPC_BRJMP = 2.U(3.W)
val NPC_CSR_EVEC   = 3.U(3.W)
val NPC_CSR_MEPC   = 4.U(3.W)
```

But I found a more scalable approach would be doing it like so:

```
// next PCs select
val NPC_WI = 2
val NPC_PCREG = 0.U(2.W)
val NPC_PLUS4 = 1.U(2.W)
val NPC_BRJMP = 2.U(2.W)
val NPC_CSR   = 3.U(2.W)
```

Where the `NPC_CSR` means "read from some CSR". And then adding a new signal called `next_pc_sel_csr_addr` which encodes which CSR to read from. 

This also in part explains the code we talked about here: https://github.com/pretis/flexpret/pull/82#discussion_r1500361646

erlingrj commented on src/main/scala/Core/control.scala:
```diff
@@ -322,23 +336,88 @@ class Control(implicit val conf: FlexpretConfiguration) extends Module
+  val exe_expire_ie_ee = io.exe_expire_ie || io.exe_expire_ee
+  val exe_expire = exe_expire_du_wu || exe_expire_ie_ee
+  
+  // At the end of DU, another DU instruction will be fetched because DU is a
+  // branch to self. We need to make sure we do not execute this DU, otherwise we
+  // will end in a loop.
+  // We do this by capturing the thread id that fecthed a DU, and marking it
+  // invalid when it otherwise would be executed.
```
I do not understand this. DU should not "branch" to itself when the DU time has already passed. The argument is an absolute time-point. So when it wakes up, fetches DU again, we have passed the absolute time and it should just continue?

magnmaeh replied:
What I wrote is based on waveform observations - I saw that for some HW thread configurations (1, 2, 3 threads), the thread executing a DU would fetch another DU when the timer had expired and execute it. Then it would just constantly loop between waking up from DU, fetching another DU, executing it, waking up, ...

The signal that indicates whether a DU has expired is called `du_expired`. This signal is per-thread (which is necessary for handling multi-threaded DUs), so it will be high if
1. thread A has executed a DU,
2. the DU has expired,
3. and thread A currently is inside the execute stage. (The time comparision happens inside the CSR module, and the CSR module is given the thread ID of the exe stage.)

The last condition is what makes it difficult. Because for, say, HWTHREADS=2, the `du_expired` will not be high before thread A has already fetched another DU (analogous to branching, it does not know the DU is the wrong instruction to fetch, because the result is not ready yet). 

For this reason, I think some kind of  "squasing/kill" mechanism for DU instructions needs to be in place. However, it is not obvious to determine which instruction to kill either. The code added is an implementation of doing this correctly for any configuration of HWTHREADS and also any schedule (in the flex case).

I have tested these mechanisms quite a lot as well for all HWTHREAD configurations. See `test_fp_wait_until`, `test_fp_delay_until`, `test_du_not_stopped_by_int`, `test_wu_stopped_by_int`. https://github.com/pretis/flexpret/blob/context-switch/programs/tests/c-tests/interrupt/interrupt.c

erlingrj replied:
Aha, I see. So there was actually a CPU bug here that you are know fixing.

magnmaeh replied:
DU/WU was not implemented for several hardware threads, so it's more of a missing feature. There used to be only one register that contained the expire time, so if several threads tried to run DU/WU at the same time, they would override each other's expire time. So this is 'unexplored' waters... But from extensive testing I think it works quite well now.

erlingrj commented on src/main/scala/Core/control.scala:
```diff
@@ -322,23 +336,88 @@ class Control(implicit val conf: FlexpretConfiguration) extends Module
+  val exe_reg_du = RegNext(dec_du.asBool)
+  val exe_reg_wu = RegNext(dec_wu.asBool)
+  
+  // Any thread may have its DU triggered at any given time. In most cases, only
+  // a single thread will have its DU triggered, but it is not unthinkable that
+  // several threads will be set off to trigger simultaneously. When 'flex' is used,
+  // the ordering of threads is configurable at run-time, so we must handle all
+  // combinations.
```
What do you mean by having its DU triggered?

magnmaeh replied:
I think I'll rewrite these comments based on what I wrote in the other question addressing this :)

erlingrj reviewed (commented):
This looks like great work Magnus! I did not inspect the tests and pin-clients very closely. I think it is critical to make sure that the changes to the CPU architecture is correct. I had a few questions there and I will go through it again after I get the responses.

magnmaeh commented on Makefile:
```diff
@@ -126,10 +129,16 @@ clean:
 	rm -rf out
+	rm -rf $(CLIENT_BUILD_DIR)
+
+	# If the hwconfig.mk file does not exist, the clean target will fail because
+	# hwconfig.mk is included in the programs/tests makefile. Therefore, create an 
+	# empty file for it to delete.
+	echo "" >> ./hwconfig.mk
```
The flow here is a bit weird. But basically what I tried to explain is that the line below (`make -C programs/tests clean`) will run the makefile in programs/tests. That makefile includes hwconfig.mk, and if hwconfig.mk does not exist it will fail.

magnmaeh commented on programs/lib/Makefrag-lib:
```diff
@@ -6,16 +6,25 @@ LIB_SOURCES += \
+
 endif
 
+LIB_AUTOGEN := $(FP_LIB_DIR)/include/flexpret_swconfig.h $(FP_LIB_DIR)/linker/flexpret_swconfig.ld
```
I looks like I missed a few of your comments and forgot to address them. I have fixed them now in the latest commit

magnmaeh reviewed (commented):
@erlingrj I noticed there were still a few issues I had not addressed. I have addressed them now and tried to close all discussions as well

erlingrj reviewed (approved):
I had another quick look and I believe it looks good. Lets get this merged :)
